subsys/net/lib/dns/mdns_responder.c

```diff
@@ -274,6 +277,96 @@ static int send_response(struct net_context *ctx,
 	return ret;
 }
 
+static const char *qtype_to_string(int qtype)
+{
+	switch (qtype) {
+	case DNS_RR_TYPE_A: return "A";
+	case DNS_RR_TYPE_CNAME: return "CNAME";
+	case DNS_RR_TYPE_PTR: return "PTR";
+	case DNS_RR_TYPE_TXT: return "TXT";
+	case DNS_RR_TYPE_AAAA: return "AAAA";
+	case DNS_RR_TYPE_SRV: return "SRV";
+	default: return "(unsupported)";
+	}
+}
+
+
+#if defined(CONFIG_MDNS_RESPONDER_DNS_SD)
+static void send_sd_response(struct net_context *ctx,
+		 struct net_pkt *pkt, union net_ip_header *ip_hdr,
+		 struct dns_msg_t *dns_msg, struct net_buf *result)
+{
+	int ret;
+	const struct dns_sd_rec *record;
+	struct dns_sd_rec filter;
+	struct sockaddr dst;
+	const struct in6_addr *addr6 = NULL;
+	const struct in_addr *addr4 = NULL;
+	char service_buf[DNS_SD_SERVICE_MAX_SIZE + 1];
+	char proto_buf[DNS_SD_PROTO_SIZE + 1];
+	char domain_buf[DNS_SD_DOMAIN_MAX_SIZE + 1];
+
+	/* This actually is used but the compiler doesn't see that */
+	ARG_UNUSED(record);
+
+#if defined(CONFIG_NET_IPV4)
+	/* Look up the local IPv4 address */
+	addr4 = net_if_ipv4_select_src_addr(net_pkt_iface(pkt),
+					   &ip_hdr->ipv4->src);
+	create_ipv4_addr(net_sin(&dst));
+	net_context_set_ipv4_ttl(ctx, 255);
+#endif /* CONFIG_NET_IPV4 */
+
+#if defined(CONFIG_NET_IPV6)
+	/* Look up the local IPv6 address */
+	addr6 = net_if_ipv6_select_src_addr(net_pkt_iface(pkt),
+					   &ip_hdr->ipv6->src);
+	create_ipv6_addr(net_sin6(&dst));
+	net_context_set_ipv6_hop_limit(ctx, 255);
+#endif /* CONFIG_NET_IPV6 */
+
+	ret = dns_sd_extract_service_proto_domain(dns_msg->msg,
+		dns_msg->msg_size, &filter, service_buf, sizeof(service_buf),
+		proto_buf, sizeof(proto_buf), domain_buf, sizeof(domain_buf));
+	if (ret < 0) {
+		NET_DBG("unable to extract service.proto.domain (%d)", ret);
+		return;
+	}
+
+	DNS_SD_FOREACH(record) {
+		/* Checks validity and then compare */
+		if (dns_sd_rec_match(record, &filter)) {
+			NET_DBG("matched query: %s.%s.%s.%s port: %u",
+				record->instance, record->service,
+				record->proto, record->domain,
+				*(record->port));
+
+			/* Construct the response */
+			ret = dns_sd_handle_ptr_query(record,
+				addr4, addr6,
+				result->data, result->size);
+			if (ret < 0) {
+				NET_DBG("dns_sd_handle_ptr_query() failed (%d)",
+					ret);
+				continue;
+			}
+
+			result->len = ret;
+
+			/* Send the response */
+			ret = net_context_sendto(ctx, result->data,
+				result->len, &dst,
+				sizeof(struct sockaddr_in6),
```

Comment:
This needs to match the proper address size. So you have to use `sizeof(struct sockaddr_in)` for IPv4 addresses.
Just have addr_len variable, and set that according to what is the actual destination address.